src/main/java/org/cryptomator/cryptofs/health/dirid/OrphanDir.java

```diff
@@ -33,5 +58,119 @@ public String toString() {
+			if (!uuid.equals(Constants.RECOVERY_DIR_ID)) {
+				throw new FileAlreadyExistsException("Directory /" + Constants.RECOVERY_DIR_NAME + " already exists, but with wrong directory id.");
+			}
+		}
+		String recoveryDirHash = cryptor.hashDirectoryId(Constants.RECOVERY_DIR_ID);
+		Path cipherRecoveryDir = dataDir.resolve(recoveryDirHash.substring(0, 2)).resolve(recoveryDirHash.substring(2)).toAbsolutePath();
+		Files.createDirectories(cipherRecoveryDir);
+
+		//create "step-parent" directory to move orphaned files to
+		String cipherStepParentDirName = convertClearToCiphertext(cryptor, clearStepParentDirName, Constants.RECOVERY_DIR_ID);
+		Path cipherStepParentDirFile = cipherRecoveryDir.resolve(cipherStepParentDirName + "/" + Constants.DIR_FILE_NAME);
+		final String stepParentUUID;
+		if (Files.exists(cipherStepParentDirFile, LinkOption.NOFOLLOW_LINKS)) {
+			stepParentUUID = Files.readString(cipherStepParentDirFile, StandardCharsets.UTF_8);
+		} else {
+			Files.createDirectory(cipherStepParentDirFile.getParent());
+			stepParentUUID = UUID.randomUUID().toString();
+			Files.writeString(cipherStepParentDirFile, stepParentUUID, StandardCharsets.UTF_8, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE);
+		}
+		String stepParentDirHash = cryptor.hashDirectoryId(stepParentUUID);
+		Path stepParentDir = dataDir.resolve(stepParentDirHash.substring(0, 2)).resolve(stepParentDirHash.substring(2)).toAbsolutePath();
+		Files.createDirectories(stepParentDir);
+		return new CryptoPathMapper.CiphertextDirectory(stepParentUUID, stepParentDir);
+	}
+
+	// visible for testing
+	void adoptOrphanedResource(Adoption adoption, CryptoPathMapper.CiphertextDirectory stepParentDir, FileNameCryptor cryptor, String longNameSuffix, MessageDigest sha1) throws IOException {
+		if (adoption.oldCipherPath.toString().endsWith(Constants.DEFLATED_FILE_SUFFIX)) {
+			var newCipherName = convertClearToCiphertext(cryptor, adoption.newClearname + longNameSuffix, stepParentDir.dirId);
+			var deflatedName = BaseEncoding.base64Url().encode(sha1.digest(newCipherName.getBytes(StandardCharsets.UTF_8))) + Constants.DEFLATED_FILE_SUFFIX;
+			Path targetPath = stepParentDir.path.resolve(deflatedName);
+			Files.move(adoption.oldCipherPath, targetPath, StandardCopyOption.ATOMIC_MOVE);
+
+			//adjust name.c9s
+			try (var fc = Files.newByteChannel(targetPath.resolve(Constants.INFLATED_FILE_NAME), StandardOpenOption.WRITE, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)) {
+				fc.write(ByteBuffer.wrap(newCipherName.getBytes(StandardCharsets.UTF_8)));
+			}
+		} else {
+			var newCipherName = convertClearToCiphertext(cryptor, adoption.newClearname, stepParentDir.dirId);
+			Path targetPath = stepParentDir.path.resolve(newCipherName);
+			Files.move(adoption.oldCipherPath, targetPath, StandardCopyOption.ATOMIC_MOVE);
+		}
+	}
+
+	// visible for testing
+	record Adoption(String newClearname, Path oldCipherPath) {}
```

Comment:
Why not just have two params in `adoptOrphanedResource`? record isn't used elsewhere, is not added to any collection or stored otherwise. Just having two params is less verbose than creating new tuples all the time.

Reply:
Fixed in 2934f3f